src/util/address.rs

```diff
@@ -18,125 +18,166 @@
         Address {
-            ty: Type::ScriptHash,
             network: network,
-            hash: Hash160::from_data(&script[..])
+            payload: Payload::ScriptHash(Hash160::from_data(&script[..])),
         }
     }
 
     /// Generates a script pubkey spending to this address
     #[inline]
     pub fn script_pubkey(&self) -> script::Script {
-        match self.ty {
-            Type::PubkeyHash => {
+        match self.payload {
+            Payload::PubkeyHash(ref hash) => {
                 script::Builder::new()
                     .push_opcode(opcodes::All::OP_DUP)
                     .push_opcode(opcodes::All::OP_HASH160)
-                    .push_slice(&self.hash[..])
+                    .push_slice(&hash[..])
                     .push_opcode(opcodes::All::OP_EQUALVERIFY)
                     .push_opcode(opcodes::All::OP_CHECKSIG)
-            }
-            Type::ScriptHash => {
+            },
+            Payload::ScriptHash(ref hash) => {
                 script::Builder::new()
                     .push_opcode(opcodes::All::OP_HASH160)
-                    .push_slice(&self.hash[..])
+                    .push_slice(&hash[..])
                     .push_opcode(opcodes::All::OP_EQUAL)
+            },
+            Payload::WitnessProgram(ref witprog) => {
+                script::Builder::new()
+                    .push_int(witprog.version as i64)
+                    .push_slice(&witprog.program[..])
             }
         }.into_script()
     }
 }
 
 impl ToString for Address {
     fn to_string(&self) -> String {
-        let mut prefixed = [0; 21];
-        prefixed[0] = match (self.network, self.ty) {
-            (Network::Bitcoin, Type::PubkeyHash) => 0,
-            (Network::Bitcoin, Type::ScriptHash) => 5,
-            (Network::Testnet, Type::PubkeyHash) => 111,
-            (Network::Testnet, Type::ScriptHash) => 196
-        };
-        prefixed[1..].copy_from_slice(&self.hash[..]);
-        base58::check_encode_slice(&prefixed[..])
+        match self.payload {
+            Payload::PubkeyHash(ref hash) => {
+                let mut prefixed = [0; 21];
+                prefixed[0] = match self.network {
+                    Network::Bitcoin => 0,
+                    Network::Testnet => 111,
+                };
+                prefixed[1..].copy_from_slice(&hash[..]);
+                base58::check_encode_slice(&prefixed[..])
+            }
+            Payload::ScriptHash(ref hash) => {
+                let mut prefixed = [0; 21];
+                prefixed[0] = match self.network {
+                    Network::Bitcoin => 5,
+                    Network::Testnet => 196,
+                };
+                prefixed[1..].copy_from_slice(&hash[..]);
+                base58::check_encode_slice(&prefixed[..])
+            }
+            Payload::WitnessProgram(ref witprog) => {
+                witprog.to_address().unwrap()
```

Comment:
This unwrap() can fail, probably should just fix upstream so taht you cant get a witprog that is invalid (ie a failing creation function).

Reply:
Good catch. Will PR upstream to hide internals.